unit_test/adapters/transforms/esdl_asset_mappers/test_esdl_asset_pipe_mapper.py

```diff
@@ -0,0 +1,138 @@
+        self.qheat_external: float = 0.0
+
+        # Arranging standard objects for the other test
+        esdl_file_path = (
+            Path(__file__).parent / ".." / ".." / ".." / ".." / "testdata" / "test1.esdl"
+        )
+        esdl_file_path = str(esdl_file_path)
+        self.esdl_object = EsdlObject(pyesdl_from_file(esdl_file_path))
+        self.mapper = EsdlAssetPipeMapper()
+
+    def test_to_entity_method(self):
+        """Test for to_entity method."""
+        pipes = self.esdl_object.get_all_assets_of_type("pipe")  # act
+        self.assertGreater(len(pipes), 0, "No pipes found in the ESDL file.")
+        self.esdl_asset = pipes[0]
+        esdl_asset = pipes[0]
+        pipe_entity = self.mapper.to_entity(esdl_asset)
+        self.assertEqual(pipe_entity.name, esdl_asset.esdl_asset.name)
+        self.assertEqual(pipe_entity.asset_id, esdl_asset.esdl_asset.id)
+        self.assertEqual(pipe_entity.connected_ports, esdl_asset.get_port_ids())  # type: ignore
+        self.assertEqual(
+            pipe_entity.length,
+            esdl_asset.get_property("length", PIPE_DEFAULTS.length)[0],  # type: ignore
+        )
+        self.assertEqual(
+            pipe_entity.diameter, self.mapper._get_diameter(esdl_asset)  # type: ignore
+        )
+        self.assertEqual(
+            pipe_entity.roughness, esdl_asset.get_property("roughness", PIPE_DEFAULTS.roughness)[0]
+        )  # type: ignore
+        self.assertEqual(
+            pipe_entity.alpha_value,  # type: ignore
+            self.mapper._get_heat_transfer_coefficient(esdl_asset),
+        )
+        self.assertEqual(
+            pipe_entity.minor_loss_coefficient,  # type: ignore
+            esdl_asset.get_property("minor_loss_coefficient", PIPE_DEFAULTS.minor_loss_coefficient)[
+                0
+            ],
+        )
+        self.assertEqual(
+            pipe_entity.external_temperature,  # type: ignore
+            esdl_asset.get_property("external_temperature", PIPE_DEFAULTS.external_temperature)[0],
+        )
+        self.assertEqual(
+            pipe_entity.qheat_external,  # type: ignore
+            esdl_asset.get_property("qheat_external", PIPE_DEFAULTS.qheat_external)[0],
+        )
+
+    def test_pipe_get_property_diameter(self):
+        """Evaluate the get property diameter method to retrieve diameters."""
+        # Arrange
+        esdl_asset_mock = Mock()
+        esdl_asset_mock.get_property.return_value = (1.0, True)
+
+        # Act
+        pass
+
+        # Assert
+        self.assertEqual(EsdlAssetPipeMapper._get_diameter(self, esdl_asset=esdl_asset_mock), 1.0)
+
+    def test_pipe_get_property_diameter_failed(self):
+        """Evaluate failure to retrieve diameter from ESDL asset."""
+        # Arrange
+        esdl_asset_mock = Mock()
+        esdl_asset_mock.get_property.return_value = (1.0, False)
+
+        # Act
+        pass
+
+        # Assert
+        with self.assertRaises(NotImplementedError):
```

Comment:
You should follow the below mentioned structure:


```suggestion
        # Act
        with self.assertRaises(ValueError) as cm:
            self.production_cluster.set_setpoints(setpoints=setpoints)

        # Assert
        self.assertIsInstance(cm.exception, ValueError)
        self.assertEqual(
            cm.exception.args[0],
            f"The setpoints {necessary_setpoints.difference(set(setpoints))} are missing.",
        )
```